kaldb/src/main/java/com/slack/kaldb/server/ElasticsearchBulkIngestAPI.java

```diff
@@ -0,0 +1,212 @@
+  private final PrometheusMeterRegistry meterRegistry;
+
+  private final KaldbConfigs.PreprocessorConfig preprocessorConfig;
+  private final DatasetMetadataStore datasetMetadataStore;
+  private final KafkaProducer<String, byte[]> kafkaProducer;
+
+  private List<DatasetMetadata> throughputSortedDatasets;
+
+  private final KaldbMetadataStoreChangeListener<DatasetMetadata> datasetListener =
+      (datasetMetadata) -> load();
+
+  private PreprocessorRateLimiter rateLimiter;
+  private Predicate<String, Trace.Span> rateLimiterPredicate;
+
+  @Override
+  protected void doStart() {
+    LOG.info("Starting ElasticsearchBulkIngestAPI service");
+    load();
+    datasetMetadataStore.addListener(datasetListener);
+    LOG.info("ElasticsearchBulkIngestAPI service started");
+  }
+
+  @Override
+  protected void doStop() {
+    LOG.info("Stopping ElasticsearchBulkIngestAPI service");
+    datasetMetadataStore.removeListener(datasetListener);
+    LOG.info("ElasticsearchBulkIngestAPI service closed");
+  }
+
+  public void load() {
+    List<DatasetMetadata> datasetMetadataList = datasetMetadataStore.listSync();
+    // only attempt to register stream processing on valid dataset configurations
+    List<DatasetMetadata> datasetMetadataToProcesses =
+        filterValidDatasetMetadata(datasetMetadataList);
+
+    this.throughputSortedDatasets = sortDatasetsOnThroughput(datasetMetadataToProcesses);
+
+    this.rateLimiter =
+        new PreprocessorRateLimiter(
+            meterRegistry,
+            preprocessorConfig.getPreprocessorInstanceCount(),
+            preprocessorConfig.getRateLimiterMaxBurstSeconds(),
+            INITIALIZE_RATE_LIMIT_WARM);
+    this.rateLimiterPredicate = rateLimiter.createRateLimiter(datasetMetadataList);
+  }
+
+  record BulkIngestResponse(int totalDocs, long failedDocs, String errorMessage) {}
+
+  public ElasticsearchBulkIngestAPI(
+      DatasetMetadataStore datasetMetadataStore,
+      KaldbConfigs.PreprocessorConfig preprocessorConfig,
+      PrometheusMeterRegistry meterRegistry) {
+    this.datasetMetadataStore = datasetMetadataStore;
+    this.preprocessorConfig = preprocessorConfig;
+    this.meterRegistry = meterRegistry;
+    this.kafkaProducer = createKafkaProducer();
+  }
+
+  /**
+   * 1. Kaldb does not support the concept of "updates". It's always an add 2. The "index" is used
+   * as the span name
+   */
+  @Blocking
+  @Post("/_bulk")
+  public HttpResponse addDocument(String bulkRequest) {
+    try {
+      Map<String, List<Trace.Span>> docs = OpenSearchRequest.parseBulkHttpRequest(bulkRequest);
+      Map<String, List<Trace.Span>> rateLimitedDocs = new HashMap<>();
+      for (Map.Entry<String, List<Trace.Span>> indexDoc : docs.entrySet()) {
+        final String index = indexDoc.getKey();
+        rateLimitedDocs.put(
+            index,
+            indexDoc.getValue().stream()
+                .filter(doc -> rateLimiterPredicate.test(index, doc))
```

Comment:
1. Say after 10 docs we hit rate limit , we will drop the rest - make it all or nothing
2. function still reads `service_name` - We pass the key as index. make use of it